In [33]:
import pandas as pd
import numpy as np
 
url = "https://raw.githubusercontent.com/mricardo89/data-mining/main/Unit-II/datasets/googleplaystore.csv"
df = pd.read_csv(url)

### Pregunta A: 
#### Ejecuten df.isnull().sum(). Identificarán que la columna Rating tiene casi 1,500 valores nulos, mientras que Type y Android Ver tienen solo 1 o 2. ¿Por qué es arquitectónicamente seguro eliminar las filas donde falta Android Ver, pero sería un error fatal eliminar las filas donde falta Rating?

In [34]:
df.isnull().sum()

App                  0
Category             0
Rating            1474
Reviews              0
Size                 0
Installs             0
Type                 1
Price                0
Content Rating       1
Genres               0
Last Updated         0
Current Ver          8
Android Ver          3
dtype: int64

##### Porque no hay comparacion a eliminar 1 celda de una columnas a eliminar 1474 celdas de otra columna, los primeros 2 procentajes son super pequeños que son hasta insignificantes, cosa que no lo es deshacernos de aproximadamente un 10% de filas.

### Pregunta B: 
#### Escriban la instrucción exacta usando dropna() pasándole el parámetro subset=['Android Ver', 'Type'] para eliminar únicamente las 3 filas que tienen esos nulos. Verifiquen que el total de registros en el DataFrame apenas disminuyó.

In [35]:
print(len(df))

print(df.isnull().sum()[['Android Ver', 'Type']])

df = df.dropna(subset=["Android Ver", "Type"])

print(len(df))

print(df.isnull().sum()[['Android Ver', 'Type']])

10841
Android Ver    3
Type           1
dtype: int64
10837
Android Ver    0
Type           0
dtype: int64


### Pregunta C: 
#### Rellenar los 1,500 nulos de Rating con el promedio global es un error, ya que las apps de la categoría FAMILY (Familia) no se califican igual que las de DATING (Citas). Utilizando la lógica de la semana pasada, escriban un bloque combinando .groupby('Category') y .transform() para rellenar los NaN de la columna Rating usando la mediana de su propia categoría.

In [36]:

df['Rating'] = df['Rating'].fillna(
    df.groupby('Category')['Rating'].transform('median')
)

print(df['Rating'].isnull().sum())


0


### Pregunta D:
#### Revisen la columna Size (Tamaño de la app). Notarán que no tiene nulos matemáticos (NaN), pero muchas filas dicen "Varies with device". Para un modelo matemático, ese texto es basura indigerible. Escriban una instrucción usando df['Size'].replace(...) para convertir exactamente esa frase de texto en un nulo matemático real usando np.nan de NumPy. ¿Por qué es mejor convertirlo a NaN en esta etapa en lugar de ponerle un tamaño falso como "0 MB"?

In [37]:
print(df['Size'])

print(df['Size'].isnull().sum())

df["Size"] = df["Size"].replace("Varies with device", np.nan)

print(df['Size'].isnull().sum())

print(df['Size'])

0                       19M
1                       14M
2                      8.7M
3                       25M
4                      2.8M
                ...        
10836                   53M
10837                  3.6M
10838                  9.5M
10839    Varies with device
10840                   19M
Name: Size, Length: 10837, dtype: object
0
1694
0         19M
1         14M
2        8.7M
3         25M
4        2.8M
         ... 
10836     53M
10837    3.6M
10838    9.5M
10839     NaN
10840     19M
Name: Size, Length: 10837, dtype: object


##### Porque podría meter un sesgo severo al crear multiples 0s, alterando la media, y sesgando la asimetría.